Add Var to complement Cmd

MichaelWedel opened the pull request:
This fixes #166.

By doing a bit more work than just introducing `Var`, `Cmd` is more flexible now and the entire process of binding the commands to the device are better defined and separated now.

A person implementing a device interface still uses `Cmd` in exactly the same way as before, minus the `regex_flags` parameter. In addition, it's now also possible to specify a callable object as the first parameter instead of a member name, which can be a free function or a lambda. Both `Cmd` and `Var` have a `bind`-method, which returns one or more `Func`-objects. Those contain all the logic that is required for determining whether a request can be processed, argument mapping, return value mapping and actually calling the associated function.

Now it's no longer necessary to modify the type of the interface to forward device methods/properties, this is done in the binding step instead.

Another side effect is that it should be massively easier to unit test this stuff, because it's all encapsulated now.

Commit:
Add Var to complement Cmd

Commit:
Use Var in example device

MichaelWedel commented:
Sorry, had forgotten some print statements, force-pushed the branch without them.

MichaelWedel commented:
Codacy is lying - I'm pretty sure that under the circumstances it's being used as a callable, it's callable. Unless of course someone specifies something that is not callable on purpose.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -86,7 +84,44 @@ def handle_accept(self):
+class BaseCommand(object):
+    member = None
+    argument_mappings = None
+    return_mapping = None
```
It might be too late to change this, but I don't like the term `mapping` here very much. It's not wrong, but a bit too general to be useful... kind of like naming a class `SomethingManager`. Something along the lines of `arg_types` and `return_filter` might be more descriptive, and more intuitive to understand what's happening when a user sees things like: 

```python
Cmd('set_target', r'^T=([-+]?[0-9]*\.?[0-9]+)$', arg_types=(float,)),
Cmd('stop', r'^H$', return_filter=lambda ret: 'T={},P={}'.format(ret[0], ret[1])),
```

But changing it now would break code that already uses these names. Do you think it would be worthwhile?

MichaelWedel replied:
I'm not sure if `return_filter` would tell me much more than `return_mapping`. To me a filter is something that removes something from something else, which is a valid use case for this, but not the only one. Maybe `transformation` was what I had in mind as a use case.

Similar for `argument_mappings` - while most of the time it might be type transformations, it could be other things as well, transforming to lower case, taking a substring, ...

Would `return_transformation` or `argument_transformations` be better? In general I think it would be okay to change it, but only if we find something that is substantially better and captures all possible uses for these two.

MikeHart85 replied:
I had completely forgotten that `argument_mappings` could be arbitrary transformation functions.

Not sure I have better terms. In a way, these are pre- and post-processing... filters for lack of a better word.. Probably best to leave it as-is in that case.

MikeHart85 replied:
`_override`, `_transform`, `_mutate`, `_process`, `_filter` (I wouldn't say this necessarily removes; consider audio/video/image filters)

MichaelWedel replied:
True, did not consider video filters and so on (that was probably the chemist speaking). I kind of like the term `transform`, because that's what it actually is right?

`argument_transforms`, `return_transform`?

MikeHart85 replied:
Transform makes me think of matrices and 3D graphics. 

The old fashioned coder in me also yearns for something much shorter... I'd probably call them `arg_map` and `ret_map`, or maybe `arg_filter` and `ret_filter`, if I was writing this. I'd leave the `s` implied in the common understanding that there can be multiple arguments but only one return value... but then, that's the C coder talking again and we're in Python land.

But this is down to personal preferences now... so I'll leave it entirely up to you. I'm happy we discussed these names a bit, but we shouldn't dwell too much on it.

MichaelWedel replied:
Well in this case I'd say we take the easy way out and leave it as it is until somebody tells us "I have no idea what you're talking about".

MikeHart85 replied:
Sounds like a plan :smile:

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -86,7 +84,44 @@ def handle_accept(self):
 
-class Cmd(object):
+class CanNotProcessException(Exception):
+    pass
```
This sort of goes back to #135... I can see how it is convenient to make ad hoc exception types where you need them to catch only that specific exception type, but is this good practice in general?

MichaelWedel replied:
It goes a bit with the comment above, I tried with different return values first, but somehow returning a mix of error code (what match essentially is) and `reply`, which is the actual result didn't really appeal to me - especially since I'm not aware we're using this "pattern" elsewhere.

On the other hand this is kind of "misusing exceptions for flow control", but only because there is no other use of `process_request` (at the moment). Otherwise I think raising an exception would be the right thing to do here - "I tried processing the request, I wasn't able to do it, maybe someone else is". I also used a special type, because that is an exception that is expected and we know how to deal with it. Other exceptions may result from actually executing the function(s) and for those we actually do not know and we catch them very broadly to forward them to the `handle_error` method.

A completely different way would be to have an additional method `can_process(request)` which just checks the regex and returns True or False. In case of actually processing the command, the request would then be run against the regex again to extract the groups. Or the `process_request` method takes a regex match object, but somehow I don't think having those in such a general method signature is particularly nice either...

I first thought this could be some sort of "Chain of Responsibility" pattern, but I think that would require a new type that is returned from the function that makes it possible to distinguish between "this is the original request" and "this is the result of the processing".

Any other ideas how to handle this?

MikeHart85 replied:
Thought about this some more. It's a bit awkward no matter how you do it. But I'm still leaning towards the tuple return.

Introducing `can_process` would give us a way to check before calling `process_request`, but then you have to use the same match in both places. And I could just call `process_request` with garbage values anyway... so it would still need to throw an exception on no match. So, in effect, we haven't simplified anything and only added code duplication and complexity.

Chain of responsibility seems like overkill for this simple situation. We could consider a simplified version, like what the `re` module does: Take parameters normally but return a `match` object that may or may not have results, and contains meta information about the nature and validity of results.

But what would our return object need to look like / what values would it contain? It would contain only two things: 

1. A boolean to indicate whether the request was accepted/processed, and 
2. A return value from processing the request (default `None` if above `False`). 

So: `(False, None)` or `(True, ret_value)`. Why create a new type for this if tuple fits the bill?

MichaelWedel replied:
From my point of view "And I could just call process_request with garbage values anyway... so it would still need to throw an exception on no match." is the key argument here. This is how I would write the function if someone told me "Write a function that processes something and return the result". I would make it raise an exception, I'm just not entirely sure which one...

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -50,14 +50,12 @@ def found_terminator(self):
+                    match = True
                     break
+                except CanNotProcessException:
+                    pass
```
Consider returning a tuple and using instead a calling convention like this:

```python
match, reply = cmd.process_request(request, self.target)
```

This would avoid the need for a custom exception and a try-except block here.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -108,16 +143,15 @@ class Cmd(object):
 
     :param target_method: Method to be called when regex matches.
     :param regex: Regex to match for method call.
-    :param regex_flags: Flags to pass ot re.compile, default is 0.
```
Since we're removing the `regex_flags` argument, we should briefly mention something about the `(?iLmsux)` feature and how to use it in the documentation above.

MichaelWedel replied:
Agreed.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -128,18 +162,71 @@ def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
 
-        return [f(a) for f, a in zip(self.argument_mappings, arguments)]
+        if not match:
+            raise CanNotProcessException()
```
As mentioned above, this could be replaced with a `return False, None`, and the successful return below with a `return True, ...`

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -128,18 +162,71 @@ def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
+        if self.read_pattern.groups != 0:
+            raise RuntimeError(
+                'Command regex for reading member \'{}\' contains '
+                'arguments.'.format(target_member))
```
While unlikely, there is a possibility that someone might want to use a capturing group not as an argument, but to backreference within the regular expression. For example, `r'([a-z]+)[0-9]+\1'` will match `'foo123foo'` and `'bar123bar'`, but not `'foo123bar'`. Treating this as an error prevents that usage.

But the same limitation exists in `Cmd`, since all capturing groups are treated as arguments.

Not something to worry about here, just to consider.

MichaelWedel replied:
Yes, I agree that we should keep that in mind for another round of improvement - in this change I'd say we have enough other things to worry about even if we don't change the behavior of `Cmd`.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -128,18 +162,71 @@ def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
+        if self.write_pattern is not None:
+            patterns.append(self.write_pattern)
+
+        return patterns
```
For brevity, this function could be reduced to:

```python
def patterns(self):
    return filter(lambda x: x is not None, [self.read_pattern, self.write_pattern])
```

Or even (if we don't care that this will also remove `0`, `False` and `''`... and we shouldn't?):

```python
def patterns(self):
    return filter(None, [self.read_pattern, self.write_pattern])
```

MichaelWedel replied:
In the light of `filter`, `map` and `reduce` being almost removed from Python 3 at some point a list comprehension would probably be "future proof" (tm).
```
def patterns(self):
    return [x for x in (self.read_pattern, self.write_pattern) if x is not None]
```

No matter which one, it would indeed make the function shorter.

MikeHart85 replied:
I'd never heard of `filter`, etc being deprecated or slated for removal. Searching for it led me to some really old (2003-2005) discussions revolving around a comment PEP 279. My feeling is that if it hasn't happened 13 years later, it's pretty safe to assume there is too much opposition to removing them and that they will stay.

I like `filter` here, especially the `None` variant. It's concise and clear, much more so than the list comprehension, which requires more... comprehending to understand. My vote is to use `filter` and prepare the pitchforks and torches in case some faceless tyrant tries to take it away from us.

MichaelWedel replied:
This is how different views can be for a simple thing...for me personally the list comprehension is easier to understand, because I haven't used `filter` in years and don't know what the it does if the first argument is `None`.

MikeHart85 replied:
Well, I'll admit it is a bit misleading... but in a good(?) way here!

It does exactly this with a `None` parameter:
```[x for x in (self.read_pattern, self.write_pattern) if x]```
(three `x`s, ugh...)

Note the missing `is not None` at the end. It simply checks for truthiness. We only care about Noneness, but it wouldn't hurt to filter the others here.

MichaelWedel replied:
While working on it I found some more problems (this needs to be `x.pattern` for one) and the logic in `__init__` was wrong. I'll push in a minute.

MikeHart85 commented on lewis/examples/simple_device/__init__.py:
```diff
@@ -39,8 +39,7 @@ class VerySimpleInterface(StreamAdapter):
     commands = {
-        Cmd('get_param', '^P$'),
-        Cmd('set_param', '^P=(.+)$'),
+        Var('param', read_regex='^P$', write_regex='^P=(.+)$')
```
Maybe keep the `Cmd` version and make the regex here match `V` instead (while still exposing `param`) just to illustrate both are possible.

MichaelWedel replied:
Good idea, will make the modifications.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -128,18 +162,71 @@ def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
+
+            if match:
+                args = self.map_arguments(match.groups())
+                return self.map_return_value(setattr(target, self.member, *args))
```
This could be a problem. Setters may need to return any number of things. Sometimes they return the new or old value. Sometimes they always return an empty string. Sometimes they don't return anything. 

You can do any of that here... but you don't want it affecting your getter.

MichaelWedel replied:
What do you propose? Having a getter mapping and a setter mapping? That would certainly solve the issue, but the question is do we want to do that?

MikeHart85 replied:
I'm not sure, I hadn't gotten that far. I just flagged it as a potential problem in the back of my mind back when I suggested `Var` in #165 (I pictured it exactly the way you implemented it here... with optional read and write regexes and all), so I knew to look for this here.

Is there anything preventing a user from splitting into two `Var`s when needed? So doing this:

```python
Var('param', read_regex='^P$', return_mapping=...),
Var('param', write_regex='^P=(.+)$', return_mapping=...),
```
Instead of a single `Var` for both? Only for the case you need custom return mappings?

If you can do this, it might not be a problem as long as it is documented.

MikeHart85 replied:
Ah, on that note... also need to be careful about calling the arguments `read_regex` and `write_regex` (and `regex` in `Cmd`).

We might want to implement `scanf` formatting as we discussed at some point. With `Cmd` it wasn't so critical because that argument wouldn't typically be named. But with `Var`, we would likely always name the argument when using it.

Should probably call them something very generic like just `read` and `write`... or maybe `read_format`, etc.

MichaelWedel replied:
There's nothing preventing the user from splitting it up into two `Var`s, so that is probably the easiest solution and we should document that in terms of some example.

That's what I meant on Slack that I wasn't happy with the parameter naming either. Maybe `read_pattern`, `write_pattern` and `pattern` are better, they wouldn't conflict with either `scanf` or `regex`.

MichaelWedel replied:
Ah there, `read_format`, that could work as well, although I would maybe find it slightly unexpected in the context of regular expressions.

MikeHart85 replied:
`*_pattern` sound great to me!

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -241,20 +328,26 @@ def _parseArguments(self, arguments):
+                if callable(getattr(self._device, member)):
+                    setattr(self, member, ForwardMethod(self._device, member))
+                else:
+                    setattr(type(self), member, ForwardProperty('_device', member, instance=self))
```
Is the `type(self)` rather than `self` intended?

MichaelWedel replied:
Yes, it's required for installing a property.

MikeHart85 replied:
Interesting, I'll have to dig a bit to understand what's going on there.

MichaelWedel replied:
This is the whole data descriptor thing...on the object it looks like an attribute, so you can never tell from the object whether it's that or a `property`. I'm not sure my understanding is good enough to explain it in a proper way...

But I do think `ForwardProperty` could be implemented in a different way, more like in `control_client`.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -86,7 +84,44 @@ def handle_accept(self):
+        if self.return_mapping is None:
+            return return_value
+
+        return self.return_mapping(return_value)
```
I don't know if this will make Codacy happy (to be fair, it has a point), but you could do this instead:
```python
    def map_return_value(self, return_value):
        if callable(self.return_mapping):
            return self.return_mapping(return_value)

        return return_value
```
Or even:
```python
    def map_return_value(self, return_value):
        if callable(self.return_mapping):
            return self.return_mapping(return_value)
        if self.return_mapping is not None:
            return self.return_mapping

        return return_value
```
This would have the side-effect of allowing unconditional, static return overrides in addition to lambda's and other callables.

MichaelWedel replied:
I like that a lot, it looks much better. But it doesn't make the naming easier :)

MikeHart85 reviewed (commented):
You still have it tagged `in progress`, but I had some feedback/comments as you're working on it.

MichaelWedel commented:
That's what the `In Progress` was intended for - discussions, because it's not finished at all.

Commit:
Address some of the problems pointed out in review

Commit:
Fix flake8 problem

MichaelWedel commented on lewis/adapters/stream.py:
```diff
@@ -128,18 +167,66 @@ def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
+                          zip(('read', 'write'), (read_pattern, write_pattern)) if
+                          pattern is not None}
+
+        if 'read' in self._patterns and self._patterns['read'].groups != 0:
```
The first part of the check wasn't there, so it would have failed for cases with only a read or write pattern. Collecting these in a dict solves the problem of `patterns` as well, no filter required there anymore, because read/write pattern are not in the dict.

MichaelWedel commented on lewis/examples/simple_device/__init__.py:
```diff
@@ -31,16 +31,25 @@ class VerySimpleInterface(StreamAdapter):
     commands = {
-        Cmd('get_param', '^P$'),
-        Cmd('set_param', '^P=(.+)$'),
+        Cmd('get_param', pattern='^V$', return_mapping=lambda x: 'The value is {}'.format(x)),
```
This now demonstrates the use of `return_mapping`.

MichaelWedel commented:
Codacy is still not happy, although the check for callable is right there. The duplication issue is somewhat valid and these simple assignments could go into the base class `__init__`.

MikeHart85 commented:
I would treat the codacy issue about non-callable as a bug on their end (or PyLint I suppose) at this point.

At most, we could replace `return_mapping = None` initialization with something callable to help it understand... but that's just silly and would be wrong.

Commit:
Change the implementation for command processing

MichaelWedel commented on lewis/adapters/stream.py:
```diff
@@ -48,20 +48,11 @@ def found_terminator(self):
-                    break
-
-            if match is None:
+            try:
```
The only thing left here that I do not like is the nested `try...except`. Otherwise `next` seems like a good solution to this problem. I don't really care that we perform the "winning match" twice, it's a small cost to pay for avoiding all the other stuff.

`process_request` still raises an exception now, but a standard one, so we do not need a new type here.

Commit:
Remove nested try/except

MichaelWedel commented on lewis/adapters/stream.py:
```diff
@@ -48,20 +48,13 @@ def found_terminator(self):
-            if match is None:
+            cmd = next((cmd for cmd in self.target.commands if cmd.can_process(request)), None)
+
+            if cmd is None:
```
Now this is much better from my point of view. Get's the first matching `Cmd` or `None` if none is found. In that case, as before, raise `RuntimeError` so it can be handled (or not). Otherwise continue processing request.

What do you think?

Commit:
Add some documentation

Commit:
Express Var in terms of Cmd

Commit:
Try more general command strategy

Commit:
Improve more general command strategy

Commit:
Fix doc problem

Commit:
Fix doc problem

Commit:
Separate Cmd and Var from processing

Commit:
Delete ForwardMethod

This is not required anymore, because callable objects
are stored directly instead of method names.

Commit:
Tests for num_args

MichaelWedel committed:
Merge branch 'master' into 166_add_var_to_stream

Commit:
Complete documentation

MikeHart85 commented on lewis/examples/simple_device/__init__.py:
```diff
@@ -31,16 +31,27 @@ class VerySimpleInterface(StreamAdapter):
 
     The device has only one parameter, which can be set to an arbitrary
-    value. The interface consists of two commands which can be invoked via telnet.
+    value. The interface consists of four commands which can be invoked via telnet.
```
Surely you mean `five`. :wink:

MikeHart85 commented on lewis/examples/simple_device/__init__.py:
```diff
@@ -31,16 +31,27 @@ class VerySimpleInterface(StreamAdapter):
+        Cmd('get_param', pattern='^V$', return_mapping='The value is {}'.format),
+        Cmd('set_param', pattern='^V=(.+)$', argument_mappings=(int,)),
+        Var('param', read_pattern='^P$', write_pattern='^P=(.+)$', doc='The only parameter.'),
+        Cmd(lambda: 4, pattern='^R$(?i)', doc='Random number')
```
The `Random` might warrant being `"Random"` instead.

MikeHart85 commented on lewis/adapters/stream.py:
```diff
@@ -86,47 +79,83 @@ def handle_accept(self):
+        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        func_args = num_args(func)
+        if self.pattern.groups != func_args:
```
You might be better off using `inspect.getcallargs` ([link](https://docs.python.org/2/library/inspect.html#inspect.getcallargs)) for this purpose. That function will allow you to *"try"* calling `func` without actually `try`ing to call it, if that makes sense.

Consider:
```python
inspect.getcallargs(func, *[None]*self.pattern.groups)
```
This will return some stuff (that you don't care about here) *IF it is possible* to call `func` with that number of arguments (without actually calling it). If it is NOT possible to call it with that amount of arguments, it will raise a `TypeError` exception.

This way, you don't need `num_args` at all from what I can tell (and the tests for it can go away as well). A shortcoming of this `num_args` implementation is that it doesn't consider optional arguments with default values, or `*args` / `**kwargs` scenarios... if `func` has anything like that, it will reject valid regexes / group counts.

MichaelWedel replied:
Thanks for that, I was considering `try...except TypeError...` but with actually calling the function, which isn't really safe.

I checked `inspect` but most likely did not read closely enough.

MikeHart85 replied:
It is sort of abusing the function... this isn't exactly what it was meant for, but it just fits so nicely.

The only reason I thought of it is because I used `getcallargs` for something else recently. It's a very useful function in general.

Commit:
Removing num_args, using inspect.getcallargs

Commit:
Fix simple_device
